Plugin and improvements

IbexLabsLuk opened the pull request:
Improved documentation and language, introduced plugin architecture and decoupled chain-specific logic from the chain module.

This enables smaller development steps and is the final design (not yet fully implememented as protocols / scrapers are missing).

Commit:
#2 - Plugin architecture re-implemented.

Commit:
#3 - Simplifying code: Added check for supported platforms, improved platform specific factories.

Commit:
#3 - Refactoring to enable plugins for platforms.

Commit:
#3 - Moved Ethereum specific indexer, subscriber and contracts into seperate module.

Commit:
#3 - Untangled ethers from the chain module. Chain module is now completely agnostic.

Commit:
#4 - Preliminary outline of consumer / producer architecture done during #3 cleanup.

Commit:
#4 - Added all the files and class skeletons required for data layer during #3 cleanup.

Commit:
#3 - Final cleanup, mostly docs

tbergmueller committed:
cli: Fix error handler

In ES6, yargs is no longer a singleton, so you need to call yargs() to
get an instance of yargs, at which point the API is identical.
https://github.com/yargs/yargs/issues/1854#issuecomment-787509517

tbergmueller committed:
readme: Complete Try it section

tbergmueller commented on packages/ethereum/src/erc20/connector.ts:
```diff
@@ -33,12 +38,12 @@ export class OiErc20 extends OiConnector {
+      plugin.log.info(`ERC20 transfer CAIP-Tags from ${caipTagFrom} (${fromAccount.toString()}), to ${caipTagTo} (${fromAccount.toString()})`);
+      plugin.log.info(`ERC20 transfer looking up addresses: ${await lookupCaipTag(caipTagFrom)} (${caipTagFrom}), ${await lookupCaipTag(caipTagTo)} (${caipTagTo})`);
     }, 
     20740100 );
```
There is a hardcoded FROM-block (this one: https://etherscan.io/block/20740100)
This makes the ethereum-module exlusively usable for Ethereum Mainnet after Sep 13, 2024. I assume this should either be a config option or the FROM-Block shall be detected from the blockchain itself?

LukOfDecen replied:
ContractFactory will get a "getDeploymentBlock" method. This was added (my fault) to test the indexer. 

Your error above could be related to the decimals and a few other minor differences between Standard erc20 and USDC

tbergmueller replied:
IMO this should be done still in this branch, to fix main asap. AFAIK on the other hand main is already broken this way - so we could also fix it independently.

Clear preference to fix it immediately though!

tbergmueller replied:
Created #12

tbergmueller reviewed (commented):
Will comment on the code itself subsequently instead of one big review. 

Currently the utils/localchain is broken, because connectors for an ERC-20 contract do not work there. Since this is an EVM-Chain and it is the same contract as the USDC contract we use for testing, this is super-strange. Currently investigating. 

So this works: `yarn run oi start --connect eip155:1/erc20:0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48`

This does not: ` yarn run oi start --connect eip155:31337/erc20:0x5FbDB2315678afecb367f032d93F642f64180aa3`

error: 
```
TypeError: Cannot read properties of undefined (reading 'subscrContract')
    at subscribeContract (/home/tbergmueller/DEV/openibex/oi_upstream/packages/chain/src/subscriber/subscribers.ts:54:67)
    at EthereumEventIndexer.subscribe (/home/tbergmueller/DEV/openibex/oi_upstream/packages/ethereum/src/indexer.ts:29:11)
    at EthereumEventIndexer.startIndexer (/home/tbergmueller/DEV/openibex/oi_upstream/packages/chain/src/indexer/indexer.ts:70:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

    ```

tbergmueller commented:
> Will comment on the code itself subsequently instead of one big review.
> 
> Currently the utils/localchain is broken, because connectors for an ERC-20 contract do not work there. Since this is an EVM-Chain and it is the same contract as the USDC contract we use for testing, this is super-strange. Currently investigating.
> 
> So this works: `yarn run oi start --connect eip155:1/erc20:0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48`
> 
> This does not: ` yarn run oi start --connect eip155:31337/erc20:0x5FbDB2315678afecb367f032d93F642f64180aa3`
> 
> error:
> 
> ```
> TypeError: Cannot read properties of undefined (reading 'subscrContract')
>     at subscribeContract (/home/tbergmueller/DEV/openibex/oi_upstream/packages/chain/src/subscriber/subscribers.ts:54:67)
>     at EthereumEventIndexer.subscribe (/home/tbergmueller/DEV/openibex/oi_upstream/packages/ethereum/src/indexer.ts:29:11)
>     at EthereumEventIndexer.startIndexer (/home/tbergmueller/DEV/openibex/oi_upstream/packages/chain/src/indexer/indexer.ts:70:10)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> 
>     ```
> ```

This is fixed in 521acf7dadf37f350ff77a76c1b1502f2463f141

tbergmueller commented on packages/ethereum/src/erc20/connector.ts:
```diff
@@ -1,29 +1,34 @@
+  constructor(assetArtifact: AssetArtifact, params: OiContractConnectorParams) {
     super(assetArtifact, params);
-    super.addPrimitive('Transfer', new OiChainTokenSupply(assetArtifact, params?.index));
+    super.addProducer('Transfer', new OiChainTokenSupply(assetArtifact, params?.index));
```
This is broken. `OiChainTokenSupply`'s constructor expects `constructor(assetArtifact: AssetArtifact, namespace = pluginNamespace, plugin = pluginName)`, hence. expects the plugin name as third parameter. You pass a setting here. 


Please fix (also, why did this not cause any problems while testing?)

tbergmueller replied:
This was apparently already broken in the old repository and not noticed so far. Since the constructor does not take any parameters which would be related to params, removing this parameter seems the right option. Done so in https://github.com/openibex/openibex/pull/10/commits/06d1741ee8c2d6582ac5355317e4e05f4ac8773e

Please confirm and resolve this conversation if this is correct

LukOfDecen replied:
Hm... I took some parameters out while documenting. Currently params are used. Definable wether caip tags get resolved or not. Was a bad idea to take the "indexed" param out, and explains why it wasnt seen by testing. (omited tests since I only "changed docs")

LukOfDecen replied:
We can keep the params out for now and gradually re-introduce them on demand. Scraper changes the connector orchestration anyways.

tbergmueller committed:
ethereum: Fix ERC20 supply connector

Has been broken since the old repository - fixing it by removing the parameter. `OiChainTokenSupply.constructor()` does not take any parameters

tbergmueller commented:
Decided together  to merge the branch (localchain and ERC-20 test from README is working). 
Subsequently, based on new main, Protocol/Index-querying as well as NPM-packages (#11 ) will be implemented.

Bugfixes have been rebased into the original commits.
